experimental/cache_size.cc

```diff
@@ -0,0 +1,70 @@
+#include <stdlib.h>
+#include <time.h>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <random>
+
+#include "../demos/asm/measurereadlatency.h"
+#include "../demos/instr.h"
+#include "../demos/utils.h"
+
+// Determines the maximum latency of reading elements of a vector of size "sz",
+// which is given as input.
+// It reads each element of the vector and measures the time
+// using MeasureReadLatency(). Later this number can be used to determine cache
+// misses and hits
+
+uint64_t AnalyzeReadingTime(int64_t sz) {
+  // Generate a random order of accesses to eliminate the impact of hw
+  // prefetchers
+  std::vector<int64_t> accesses(sz);
+  for (int64_t i = 0; i < sz; i++) {
+    accesses.push_back(i);
+  }
+  std::mt19937 f(std::random_device());
+  std::shuffle(accesses.begin(), accesses.end(), f);
+
+  std::vector<char> buf(sz);
+
+  // TODO: not sure about the size of chuncks and its impact on prefetchers
+  for (int64_t i : accesses) {
+    ForceRead(&buf[i]);
+  }
+
+  // Read each element in the same random order and keep track of the slowest
+  // read.
+  uint64_t max_read_latency = std::numeric_limits<uint64_t>::min();
+  for (int64_t i : accesses) {
+    max_read_latency = std::max(max_read_latency, MeasureReadLatency(&buf[i]));
+  }
+
+  return max_read_latency;
+}
+
+int main() {
+  int64_t max_size = 16 * 1024 * 1024;
```

Comment:
(optional) FWIW all three of these could be static constexpr variables to make it clearer they are constants:

```c++
static constexpr int64_t kMaxSize = 16 * 1024 * 1024;
static constexpr int64_t kMinSize = 1024;
static constexpr int iterations = 30;
```

I would also move them directly above the first for loop, following the advice at https://google.github.io/styleguide/cppguide.html#Local_Variables: " We encourage you to declare them in as local a scope as possible, and as close to the first use as possible."

Making them local variables instead of global was a great call (that's the first thing they mention!), the actual exact location is much more of a nit-pick and doesn't matter as much.

Reply:
okay.